File: problems/153_find-minimum-in-rotated-sorted-array/main.cpp
#include <iostream>
#include <vector>

class Solution
{
public:
    int findMin(std::vector<int> &nums)
    {
        int l = 0, r = (int)nums.size() - 1;
        while (l < r)
        {
            int mid = l + (r - l) / 2;
            if (nums[mid] > nums[r])
            {
                l = mid + 1; // min is to the right
            }
            else
            {
                r = mid; // min is at mid or to the left
            }
        }
        return nums[l];
    }
};

int main()
{
    Solution solution;
    std::vector<std::pair<std::vector<int>, int>> samples = {
        {{3, 4, 5, 1, 2}, 1},
        {{4, 5, 6, 7, 0, 1, 2}, 0},
        {{11, 13, 15, 17}, 11},
    };

    for (const auto &[nums, expected] : samples)
    {
        auto input = nums;
        int result = solution.findMin(input);

        std::cout << "input: nums = [";
        for (size_t i = 0; i < nums.size(); ++i)
        {
            std::cout << nums[i] << (i + 1 < nums.size() ? "," : "");
        }
        std::cout << "]"
                  << ", placeholder result: " << result
                  << ", expected: " << expected << '\n';
    }

    return 0;
}

File: problems/153_find-minimum-in-rotated-sorted-array/main.rs
struct Solution;

impl Solution {
    pub fn find_min(nums: Vec<i32>) -> i32 {
        let mut l = 0;
        let mut r = nums.len() as i32 - 1;
        while l < r {
            let mid = l + (r - l) / 2;
            if nums[mid as usize] > nums[r as usize] {
                l = mid + 1;
            } else {
                r = mid;
            }
        }
        nums[l as usize]
    }
}

fn main() {
    let samples: Vec<(Vec<i32>, i32)> = vec![
        (vec![3, 4, 5, 1, 2], 1),
        (vec![4, 5, 6, 7, 0, 1, 2], 0),
        (vec![11, 13, 15, 17], 11),
    ];

    for (nums, expected) in samples {
        let result = Solution::find_min(nums.clone());
        println!(
            "input: nums = {:?}, placeholder result: {}, expected: {}",
            nums, result, expected
        );
    }
}
